### PR TITLE
phpstan fixes part 2, ECS fixes; OPSRC-405

### PR DESCRIPTION
**Pull request by piotrleo**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | applies ECS and phpstan fixes, OPSRC-405
| License         | MIT

### CODE REVIEW

**Comment by kptLucek** [comment]
.

File: src/Twig/Runtime/RenderLinkRuntime.php [hunk 1]
```diff
@@ -59,9 +61,10 @@ public function getLinkForCode(
         string $code,
         array $options = []
     ): string {
+        /** @var PageInterface|null $page */
         $page = $this->pageRepository->findOneEnabledByCode($code, $this->localeContext->getLocaleCode());
-        if (!$page) {
-            return $options['notFoundMessage'] ?? '';
+        if (null === $page) {
+            throw new NotFoundHttpException('Page for code ' . $code . ' not found');
```
[the remark this answers is not in the document]
Comment (piotrleo) [reply src/Twig/Runtime/RenderLinkRuntime.php:67]: I don't get it, WDYM by "put argument into quotes"?

File: src/Controller/PageController.php [hunk 2]
```diff
@@ -90,15 +98,21 @@ private function resolveImage(PageInterface $page): void
...
+        $fileContents = file_get_contents($file->getPathname());
+        assert(is_string($fileContents));
+        $base64Content = base64_encode($fileContents);
         $path = 'data:' . $file->getMimeType() . ';base64, ' . $base64Content;
-
         $image->setPath($path);
-        $page->getTranslation()->setImage($image);
+        /** @var PageTranslationInterface $pageTranslationInterface */
+        $pageTranslationInterface = $page->getTranslation();
+        $pageTranslationInterface->setImage($image);
```
[the remark this answers is not in the document]
Comment (piotrleo) [reply src/Controller/PageController.php:112]: I don't know where is it copy-pasted in code fragment you highlighted. But I've done some refactoring to remove duplicates.

File: phpstan.neon [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+parameters:
+    reportUnmatchedIgnoredErrors: false
+    checkMissingIterableValueType: false
+    checkGenericClassInNonGenericObjectType: false
+    excludes_analyse:
+        # Sitemap dependent providers
+        - 'src/SitemapProvider'
+        - 'src/Importer/AbstractImporter'
+        - 'tests/Application/config/bootstrap.php'
+        - 'tests/Fixture/PageFixtureTest.php'
```
[the remark this answers is not in the document]
Comment (piotrleo) [reply phpstan.neon:12]: 'tests/Fixture/PageFixtureTest.php' depends on Behat, which is not obligatory for this plugin

File: src/Twig/Extension/RenderProductPagesExtension.php [hunk 4]
```diff
@@ -10,58 +10,87 @@
...
-
         $parsedDate = null;
-        if (!empty($date)) {
-            $parsedDate = new \DateTimeImmutable($date);
+        if (null !== $date) {
+            $parsedDate = new DateTimeImmutable($date);
         }
-
+        /** @var PageRepositoryInterface $pageRepository */
+        $pageRepository = $this->entityManager->getRepository(Page::class);
```
Comment (er1z) [inline src/Twig/Extension/RenderProductPagesExtension.php:75]: Inject repository as a constructor argument.

File: src/Twig/Runtime/RenderLinkRuntime.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (er1z) [reply src/Twig/Runtime/RenderLinkRuntime.php:67]: `'Page for code "' . $code . '" not found'`

**Review by er1z — CHANGES_REQUESTED** [review]

File: src/CustomResourceController.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+use BitBag\SyliusCmsPlugin\Controller\Helper\FormErrorsFlashHelperInterface;
+use BitBag\SyliusCmsPlugin\Resolver\ResourceResolver;
+use BitBag\SyliusCmsPlugin\Resolver\ResourceResolverInterface;
+use Sylius\Bundle\ResourceBundle\Controller\RequestConfiguration;
+use Sylius\Bundle\ResourceBundle\Controller\ResourceController;
+use Sylius\Component\Resource\Model\ResourceInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+abstract class CustomResourceController extends ResourceController
```
[the remark this answers is not in the document]
Comment (piotrleo) [reply src/CustomResourceController.php:16]: Actually I had no idea how to name it properly. Any suggestions would really help me

File: src/Controller/BlockController.php [hunk 6]
```diff
@@ -38,6 +41,8 @@ public function renderBlockAction(Request $request): Response
         $this->eventDispatcher->dispatch(ResourceActions::SHOW, $configuration, $block);
 
         if (!$configuration->isHtmlRequest()) {
+            Assert::true(null !== $this->viewHandler, 'View handler is null');
```
Comment (er1z) [inline src/Controller/BlockController.php:44]: If you use Webmozart's library, you can check `Assert::isNotNull` which is better in terms of readability.

File: src/Controller/BlockController.php [hunk 7]
```diff
@@ -70,6 +75,8 @@ public function previewAction(Request $request): Response
         $block->setCurrentLocale($request->get('_locale', $defaultLocale));
 
         if (!$configuration->isHtmlRequest()) {
+            Assert::true(null !== $this->viewHandler, 'View handler is null');
```
Comment (er1z) [inline src/Controller/BlockController.php:78]: As above.

File: src/Importer/BlockImporter.php [hunk 8]
```diff
@@ -83,7 +83,7 @@ public function import(array $row): void
 
         $this->validateResource($block, ['bitbag']);
 
-        $block->getId() ?: $this->entityManager->persist($block);
+        $this->entityManager->persist($block);
```
Comment (er1z) [inline src/Importer/BlockImporter.php:86]: Why not `Repository::add`?

File: src/Importer/MediaImporter.php [hunk 9]
```diff
@@ -77,7 +77,7 @@ public function import(array $row): void
 
         $this->validateResource($media, ['bitbag']);
 
-        $media->getId() ?: $this->entityManager->persist($media);
+        $this->entityManager->persist($media);
```
Comment (er1z) [inline src/Importer/MediaImporter.php:80]: Why not `Repository::add`?

File: src/Sorter/SectionsSorter.php [hunk 10]
```diff
@@ -20,8 +21,10 @@ public function sortBySections(array $pages): array
 
         /** @var PageInterface $page */
         foreach ($pages as $page) {
+            Assert::isIterable($page->getSections());
             foreach ($page->getSections() as $section) {
                 $sectionCode = $section->getCode();
+                Assert::notNull($sectionCode);
```
Comment (er1z) [inline src/Sorter/SectionsSorter.php:27]: Besides, why don't break this into private methods?

File: src/Twig/Parser/ContentParser.php [hunk 11]
```diff
@@ -70,15 +73,26 @@ private function getFunctionArguments(string $functionName, string $input): ?arr
...
+    ): string {
         Assert::keyExists($functions, $functionName, sprintf('Function %s does not exist!', $functionName));
         /** @var \Twig_Function $function */
         $function = $functions[$functionName];
         $callable = $function->getCallable();
+        Assert::isArray($callable, sprintf('Function with name "%s" is not callable', $functionName));
         $extension = $callable[0];
         $extensionMethod = $callable[1];
+        $callback = [$extension, $extensionMethod];
+        if (!is_callable($callback)) {
```
Comment (er1z) [inline src/Twig/Parser/ContentParser.php:92]: `Assert` class contains `isCallable` method.

File: tests/Behat/Resources/services/pages.yml [hunk 12]
```diff
@@ -1,3 +1,8 @@
 imports:
     - { resource: "pages/admin.yml" }
     - { resource: "pages/shop.yml" }
+services:
+    bitbag_sylius_cms_plugin.behat.page:
+        class: FriendsOfBehat\PageObjectExtension\Page\Page
+        arguments:
+          [ 'behat.mink.default_session', 'behat.mink.parameters' ]
```
Comment (er1z) [inline tests/Behat/Resources/services/pages.yml:8]: New line at the end of file, please.

File: src/Controller/MediaController.php [hunk 13]
```diff
@@ -108,16 +114,35 @@ public function previewAction(Request $request): Response
         ]);
     }
 
-    private function resolveFile(MediaInterface $media): void
+    private function setResourceMediaPathIfExists(MediaInterface $media): void
```
Comment (er1z) [inline src/Controller/MediaController.php:117]: What does it do?

File: src/Controller/ResourceDataProcessingTrait.php [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        return $this->requestConfigurationFactory->create($this->metadata, $request);
+    }
+
+    protected function setResourceMediaPath(MediaInterface $media): void
+    {
+        /** @var string|null $mediaPath */
+        $mediaPath = $media->getPath();
+        Assert::notNull($mediaPath, 'Media path is null');
+        Assert::string($this->getParameter('sylius_core.public_dir'));
+        $file = $media->getFile() ?? new File($this->getParameter('sylius_core.public_dir') . '/' . $media->getPath());
```
Comment (er1z) [inline src/Controller/ResourceDataProcessingTrait.php:40]: Extract it via CacheResolver; the image can be locally stored file or cloud one, so if you decide to fetch it this way: only the further will be working.

**Review by er1z — CHANGES_REQUESTED** [review]

File: src/Twig/Parser/ContentParser.php (hunk 11 above)
Comment (piotrleo) [reply src/Twig/Parser/ContentParser.php:92]: Right, I've had it done, but reverted and forgot about it

File: src/Controller/BlockController.php (hunk 6 above)
Comment (piotrleo) [reply src/Controller/BlockController.php:44]: I know, but NotNull gave me an error in phpstan for Sylius 1.8, coz the change to ViewHandler has been made when Sylius was updated to 1.9. Earlier the ViewHandler was not nullable.

File: src/Controller/BlockController.php (hunk 7 above)
Comment (piotrleo) [reply src/Controller/BlockController.php:78]: As above, these are only two cases where I wrote it like this.

File: tests/Behat/Resources/services/pages.yml (hunk 12 above)
Comment (piotrleo) [reply tests/Behat/Resources/services/pages.yml:8]: Sure :)

File: src/Controller/MediaController.php (hunk 13 above)
Comment (piotrleo) [reply src/Controller/MediaController.php]: Yeah, it was stupid function, I've just removed it :)

File: src/Importer/MediaImporter.php (hunk 9 above)
Comment (piotrleo) [reply src/Importer/MediaImporter.php]: Done

File: src/Twig/Parser/ContentParser.php (hunk 11 above)
Comment (piotrleo) [reply src/Twig/Parser/ContentParser.php]: Done

File: tests/Behat/Resources/services/pages.yml (hunk 12 above)
Comment (piotrleo) [reply tests/Behat/Resources/services/pages.yml]: Done

File: src/Importer/BlockImporter.php (hunk 8 above)
Comment (piotrleo) [reply src/Importer/BlockImporter.php]: Done

File: src/Sorter/SectionsSorter.php (hunk 10 above)
Comment (piotrleo) [reply src/Sorter/SectionsSorter.php]: Done

File: src/Controller/ResourceDataProcessingTrait.php (hunk 14 above)
Comment (piotrleo) [reply src/Controller/ResourceDataProcessingTrait.php]: There are not only images, I left it like it was before

File: src/Controller/ResourceDataProcessingTrait.php [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        Assert::string($fileContents);
+        $this->setFileContentsAsMediaPath($media, $fileContents);
+    }
+
+    private function setFileContentsAsMediaPath(MediaInterface $media, string $fileContents): void
+    {
+        $base64Content = base64_encode($fileContents);
+        $path = 'data:' . $media->getMimeType() . ';base64, ' . $base64Content;
+        $media->setPath($path);
+    }
```
Comment (er1z) [inline src/Controller/ResourceDataProcessingTrait.php:101]: Isn't the amount of logic too large for a single trait…?

File: src/Twig/Parser/ContentParser.php [hunk 16]
```diff
@@ -70,15 +72,24 @@ private function getFunctionArguments(string $functionName, string $input): ?arr
...
+    private function callFunction(
+        array $functions,
+        string $functionName,
+        array $arguments
+    ): string {
         Assert::keyExists($functions, $functionName, sprintf('Function %s does not exist!', $functionName));
         /** @var \Twig_Function $function */
         $function = $functions[$functionName];
         $callable = $function->getCallable();
+        Assert::isArray($callable, sprintf('Function with name "%s" is not callable', $functionName));
```
[the remark this answers is not in the document]
Comment (piotrleo) [reply src/Twig/Parser/ContentParser.php:87]: Wdym by redundant? Get callable can also be null and stan gives following errors:
```
  88     Cannot access offset 0 on (callable(): mixed)|null.  
  89     Cannot access offset 1 on (callable(): mixed)|null.  
```

File: src/Controller/ResourceDataProcessingTrait.php (hunk 15 above)
Comment (piotrleo) [reply src/Controller/ResourceDataProcessingTrait.php:74]: You're right, done

File: src/Twig/Parser/ContentParser.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (piotrleo) [reply src/Twig/Parser/ContentParser.php:83]: I'm responding in here to the first comment in this CR, coz I have no possibilty to respond to that. What I meant in commit message was that I added custom Assert error messages. I will remove redundant ones, which are generic like `'View handler is null'`

[the remark this answers is not in the document]
Comment (piotrleo) [reply src/Twig/Parser/ContentParser.php:83]: Done

File: src/Controller/MediaController.php [hunk 17]
```diff
@@ -11,55 +11,62 @@
...
-            return new Response();
-        }
+        Assert::notNull($media);
 
         $this->eventDispatcher->dispatch(ResourceActions::SHOW, $configuration, $media);
 
+        /** @var string|null $mediaPath */
+        $mediaPath = $media->getPath();
+        Assert::notNull($mediaPath);
+        Assert::string($this->getParameter('sylius_core.public_dir'), 'sylius_core.public_dir is not string');
```
Comment (er1z) [inline src/Controller/MediaController.php:69]: Do we still deal with locally stored asset?

File: src/Controller/ResourceDataProcessingTrait.php [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+        $fileContents = file_get_contents($resolvedPath);
+        Assert::string($fileContents);
+        $this->setFileContentsAsMediaPath($media, $fileContents);
+    }
+
+    private function setPathForNonImageFile(MediaInterface $media): void
+    {
+        Assert::string($media->getPath());
+        Assert::string($this->getParameter('sylius_core.public_dir'));
+        $file = new File($this->getParameter('sylius_core.public_dir') . $media->getPath());
```
Comment (er1z) [inline src/Controller/ResourceDataProcessingTrait.php:63]: Doesn't `MediaController` contain similar code?

File: src/Twig/Parser/ContentParser.php (hunk 16 above)
[the remark this answers is not in the document]
Comment (er1z) [reply src/Twig/Parser/ContentParser.php:83]: I meant that passing `$callable` array directly to `isCallable` and thus to the callback is solving the same problem with a less amount of code. :)

File: src/Controller/MediaController.php [hunk 19]
```diff
@@ -11,56 +11,56 @@
...
 
 final class MediaController extends ResourceController
 {
-    public function renderMediaAction(Request $request): Response
-    {
-        $configuration = $this->requestConfigurationFactory->create($this->metadata, $request);
+    use ResourceDataProcessingTrait;
 
-        $this->isGrantedOr403($configuration, ResourceActions::SHOW);
+    use CommonSetterDependencyInjectionsTrait;
```
Comment (er1z) [inline src/Controller/MediaController.php:30]: How about renaming to more context-specific?

File: src/Resources/assets/admin/js/index.js [hunk 20]
```diff
@@ -0,0 +1,17 @@
+import {HandleCsvUpload, HandleSlugUpdate, HandlePreview, HandleAutoComplete} from './bitbag';
+
+if (document.querySelector('[data-bb-target="cms-import"]')) {
```
Comment (er1z) [inline src/Resources/assets/admin/js/index.js:3]: Is it intentional that a few scenarios can happen at once?

File: src/Resources/public/build/manifest.json [hunk 21]
```diff
@@ -0,0 +1,6 @@
+{
+  "public/build/bitbag-cms-shop.css": "/public/build/bitbag-cms-shop.css",
+  "public/build/bitbag-cms-shop.js": "/public/build/bitbag-cms-shop.js",
+  "public/build/bitbag-cms-admin.css": "/public/build/bitbag-cms-admin.css",
+  "public/build/bitbag-cms-admin.js": "/public/build/bitbag-cms-admin.js"
+}
```
Comment (er1z) [inline src/Resources/public/build/manifest.json:6]: New line at the end?

File: tests/Application/assets/shop/entry.js [hunk 22]
```diff
@@ -0,0 +1 @@
+import 'sylius/bundle/ShopBundle/Resources/private/entry';
```
Comment (er1z) [inline tests/Application/assets/shop/entry.js:1]: Empty line missing.

File: src/Resources/assets/admin/js/index.js (hunk 20 above)
Comment (piotrleo) [reply src/Resources/assets/admin/js/index.js:3]: It's actually change from master, I'll leave it without changes but we can take up the topic on os-ongoing

File: src/Resources/public/build/manifest.json (hunk 21 above)
Comment (piotrleo) [reply src/Resources/public/build/manifest.json:6]: Done

File: tests/Application/assets/shop/entry.js (hunk 22 above)
Comment (piotrleo) [reply tests/Application/assets/shop/entry.js:1]: Done

File: src/Controller/MediaController.php (hunk 19 above)
Comment (piotrleo) [reply src/Controller/MediaController.php:30]: Done

File: src/Controller/MediaController.php [hunk 23]
```diff
@@ -76,29 +76,28 @@ public function downloadMediaAction(Request $request): Response
...
-            $this->repository->find($request->get('id')) :
-            $this->factory->createNew();
-        $form = $this->resourceFormFactory->create($configuration, $media);
+        $media = $this->getResourceInterface($request);
+        $form = $this->getFormForResource($configuration, $media);
         $mediaTemplate = null;
 
         $form->handleRequest($request);
 
         if ($form->isValid()) {
```
Comment (er1z) [inline src/Controller/MediaController.php:89]: Can this block be extracted to private method? Nesting too much causes code to be less readable.

**Review by er1z — APPROVED** [review]

**Pull request merged**